src/main/java/graphql/execution/DeferSupport.java

```diff
@@ -0,0 +1,113 @@
+package graphql.execution;
+
+import graphql.Directives;
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.GraphQLError;
+import graphql.Internal;
+import graphql.language.Field;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+/**
+ * This provides support for @defer directives on fields that mean that results will be sent AFTER
+ * the main result is sent via a Publisher stream.
+ */
+@Internal
+public class DeferSupport implements Publisher<ExecutionResult> {
+
+    private final Deque<Supplier<CompletableFuture<ExecutionResult>>> deferredCalls = new ConcurrentLinkedDeque<>();
+    private final List<GraphQLError> errorsEncountered = new ArrayList<>();
+    private final AtomicBoolean deferDetected = new AtomicBoolean(false);
+
+    public boolean checkForDeferDirective(List<Field> currentField) {
+        for (Field field : currentField) {
+            if (field.getDirective(Directives.DeferDirective.getName()) != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super ExecutionResult> subscriber) {
+        subscriber.onSubscribe(emptySubscription());
+        boolean ok = drainDeferredCalls(subscriber);
+        if (ok) {
+            subscriber.onComplete();
+        }
+    }
+
+    private boolean drainDeferredCalls(Subscriber<? super ExecutionResult> subscriber) {
+        while (!deferredCalls.isEmpty()) {
+            Supplier<CompletableFuture<ExecutionResult>> deferredExecution = deferredCalls.pop();
+            try {
+                CompletableFuture<ExecutionResult> future = deferredExecution.get();
+                //
+                // this becomes a blocking call on the deferred resolveField value.  This and the use of the queue
+                // means we will resolve them in "encountered" order.  We could do it async and hence in any order
+                //
+                ExecutionResult executionResult = future.get();
+                executionResult = addAnyErrorsEncountered(executionResult);
+                subscriber.onNext(executionResult);
+            } catch (Exception e) {
+                subscriber.onError(e);
+                return false;
+            }
+        }
+        return true;
+    }
```

Comment:
key with the deferredCalls queue is that it executes in "encountered order".  So we KNOW that any inner calls that might cause deferred fields will be simply added to the queue while executing the parent path.

so

      parent {
           child1 @defer {
                   x
                   childChild1 @defer {
                       ...
                   childChild2 @ defer {
                        ...
                   }
          child2 @defer {
                  ....


will result in a queue of [child1, child2, childChild1, childChild2 ] and the latter grand children will only be added when the call to child1 is made and they go to the end of the queue.